### PR TITLE
JDK-8283865: riscv: Break down -XX:+UseRVB into seperate options for each bitmanip extension

### PR DESCRIPTION
**Pull request by feilongjiang**
Currently openjdk riscv supports RISC-V bitmanip extension as a bundle while spec provides four individual extensions: Zb[abcs]\[1].

According to the spec, we need to break down `UseRVB` into two individual options `UseZba` and `UseZbb` to enable or disable Zba and Zbb respectively (openjdk riscv only supports Zba and Zbb for now).

Since multi-letter extensions representation in the ISA bitmap is still not determined \[2]\[3], availability for those extensions could not be queried from HWCAP. Feature detection of Zba and Zbb was removed temporarily.

 Linux RISCV64 release hotspot/jdk tier1 tests are  passed on QEMU with following options:
- [x] +UseZba && +UseZbb
- [x] +UseZba && -UseZbb
- [x] -UseZba && +UseZbb

\[1]: https://github.com/riscv/riscv-bitmanip/releases/download/1.0.0/bitmanip-1.0.0-38-g865e7a7.pdf
\[2]: http://lists.infradead.org/pipermail/linux-riscv/2021-November/010250.html
\[3]: http://lists.infradead.org/pipermail/linux-riscv/2021-November/010252.html

---------
### Progress
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue
- [x] Change must be properly reviewed

### Issue
 * [JDK-8283865](https://bugs.openjdk.java.net/browse/JDK-8283865): riscv: Break down -XX:+UseRVB into seperate options for each bitmanip extension

### Reviewers
 * [Fei Yang](https://openjdk.java.net/census#fyang) (@RealFYang - Committer) ⚠️ Review applies to da7f419b9d99836b0d48c511397bab1fb9330425
 * [Aleksey Shipilev](https://openjdk.java.net/census#shade) (@shipilev - **Reviewer**) ⚠️ Review applies to 08cd08eb41f33c4b78a959c000f60fd09a20de23

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.java.net/jdk pull/8032/head:pull/8032` \
`$ git checkout pull/8032`

Update a local copy of the PR: \
`$ git checkout pull/8032` \
`$ git pull https://git.openjdk.java.net/jdk pull/8032/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 8032`

View PR using the GUI difftool: \
`$ git pr show -t 8032`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.java.net/jdk/pull/8032.diff">https://git.openjdk.java.net/jdk/pull/8032.diff</a>

</details>

### CODE REVIEW

File: src/hotspot/cpu/riscv/macroAssembler_riscv.cpp [hunk 1]
```diff
@@ -1521,7 +1521,7 @@ void MacroAssembler::revb_h_h_u(Register Rd, Register Rs, Register tmp) {
 // reverse bytes in halfwords in lower 32 bits and zero-extend
 // Rd[31:0] = Rs[23:16] Rs[31:24] Rs[7:0] Rs[15:8] (zero-extend to 64 bits)
 void MacroAssembler::revb_h_w_u(Register Rd, Register Rs, Register tmp1, Register tmp2) {
-  if (UseRVB) {
+  if (UseZba && UseZbb) {
```
[the remark this answers is not in the document]
Comment (feilongjiang) [reply src/hotspot/cpu/riscv/macroAssembler_riscv.cpp]: Yes indeed. Actually only `zext_w` belongs to Zba, I just use `zero_extend` instead.

**Review by RealFYang — APPROVED** [review]
Changes looks good to me. (Not a JDK Reviewer)

**Comment by feilongjiang** [comment]
hotspot/jdk tier1 tests on QEMU with options `-XX:+UseZba -XX:+UseZbb` still good after changes.

**Comment by shipilev** [comment]
Sorry, I probably misunderstand the background and the consequences here.

This patch effectively disables B-extensions unless user specifically enables the subextensions with `+UseZba` and `+UseZbb`? Is there a hardware that supports B-extension only partially? I.e. do we care about this, at expense of disabling the B-extension by default?

It looks to me we need to postpone this patch until we can detect Zba and Zbb availability automatically.

**Comment by RealFYang** [comment]
> Sorry, I probably misunderstand the background and the consequences here.
> 
> This patch effectively disables B-extensions unless user specifically enables the subextensions with `+UseZba` and `+UseZbb`? Is there a hardware that supports B-extension only partially? I.e. do we care about this, at expense of disabling the B-extension by default?
> 
> It looks to me we need to postpone this patch until we can detect Zba and Zbb availability automatically.

Let me try to clarify:

This patch effectively disables B-extensions unless user specifically enables the subextensions with +UseZba and +UseZbb?

-- That's correct. This is also true even before this change. 
    Support for those new ISA-features are currently treated experimental and disabled by default. 
    Main reason is the lack of hardware supporting those ISA-extensions. 
    In fact, those ISA-extensions are rather new and I believe hardware supporting those ISA-extensions are not in the market for now.

Is there a hardware that supports B-extension only partially? I.e. do we care about this, at expense of disabling the B-extension by default?

--  No. Actually, we can only test this with QEMU.  

It looks to me we need to postpone this patch until we can detect Zba and Zbb availability automatically. 

-- In fact this patch fixes two things here:

    1. The bitmanipulation ISA extensions will be ratified as individual small extension packages instead of a large B-extension. 
        Software tools should follow this spec. One typical example is QEMU [1]. 
        Previously, QEMU also support this a bundle, so you do: qemu-riscv64 -cpu rv64,x-b=true
        As the spec is frozen now, that x-b=true option is no there anymore. Instead, now you do: qemu-riscv64 -cpu rv64,zba=true,zbb=true,zbc=true,zbs=true
        I think we should do similar thing for JVM breaking down the -XX:-UseRVB option into separate options for each Zb* extension.
 
    2. The current way for detecting those ISA-extensions is wrong. I think this was once affected by some early draft spec. 
         This patch fixes this by removing the wrong way for detection and leaves a FIXME there. 
         I think that's fine to me before the correct way is settled down on Linux/RISCV64. 

[1] https://patchwork.kernel.org/project/qemu-devel/patch/20210911140016.834071-5-philipp.tomsich@vrull.eu/

**Review by shipilev — APPROVED** [review]
All right, thanks for explanations. This PR makes sense then.

**Comment by feilongjiang** [comment]
@RealFYang @shipilev -- Thanks for the review! The Windows failure is some unrelated environment setup problem, which is not caused by this patch.

**Pull request closed without merging**